src/domain/common/lifecycle/lifecycle.service.ts

```diff
@@ -82,14 +92,20 @@ export class LifecycleService {
     this.logger.verbose?.(
       `Lifecycle (id: ${
         lifecycle.id
-      }) event '${eventName}: from state '${startState}' to state '${machineService.state.value.toString()}', parentID: ${parentID}`,
+      }) event '${eventName}: from state '${restoredState.value.toString()}' to state '${machineService.state.value.toString()}', parentID: ${parentID}`,
       LogContext.LIFECYCLE
     );
 
-    const newStateStr = JSON.stringify(machineService.state);
+    const stateToHydrate = interpretedLifecycle.state;
+    // Note: https://github.com/statelyai/xstate/discussions/1757
+    // restoring state has the hydrated actions, which unless removed will be executed again
+    stateToHydrate.actions = [];
```

Comment:
Ok, reading the discussion above I see what was the issue, and the solution (which is the only solution for us now).
A properly defined architectural solution restoring the whole machine without changing it can be done with CQRS, but we are way off it for now. Alternatively, XState v5, if they decide to implement it.

Reply:
The alternative design would be to always have a machine active for every entity, and this would be a very interesting approach. 

However even in that case we would still need to be able to hydrate / rehydrate in a reliable way. 

My take is that this is simply a flaw in XState, basically implying that the kind of scenario we are using it for is not what it is designed for (long running server processes).